### PR TITLE
Group line items properly

### PR DESCRIPTION
**Pull request by Jaimies**
Not very polished, but that's what I have so far.

I'll clean it up tomorrow.

Btw, psalm is freaking out, because I'm using enums, but afaik that can be fixed by updating it.

### CODE REVIEW

**Comment by Jaimies** [comment]
I should probably also add tests for the new `Segment`s. Though I also have an idea of making a base abstract class for segments, because they are all so incredibly similar. I that case, I would most likely only have to test the root class, not every single subclass. But that's for later

**Comment by Jaimies** [comment]
Btw, I just realised that using enums makes this code fail on PHP 8.0, which I'm not sure if the library can afford

**Comment by Jaimies** [comment]
What do you think, should we increase min PHP version to 8.1? I mean, the usage of enums there is not critical, but it does make the code more logical and readable.

**Comment by Jaimies** [comment]
Added tests

**Comment by Jaimies** [comment]
Btw, the segments inside Line Items should also be grouped like the normal segments, as follows:

```
[
    ...
    'LIN' => [
        'QTY' => [
            '21' => QTYQuantity(...),
            '23' => QTYQuantity(...),
        ],
        ...
    ],
    ...
]
```

Because I'm pretty sure, that a single item may contain multiple quantities of different types, but I should check.

**Comment by Jaimies** [comment]
Got that done, but now the question is how to correctly determine the end of the "line items" section.

> Btw, the segments inside Line Items should also be grouped like the normal segments, as follows:
> 
> ```
> [
>     ...
>     'LIN' => [
>         'QTY' => [
>             '21' => QTYQuantity(...),
>             '23' => QTYQuantity(...),
>         ],
>         ...
>     ],
>     ...
> ]
> ```
> 
> Because I'm pretty sure, that a single item may contain multiple quantities of different types, but I should check.

**Comment by Jaimies** [comment]
From what I gathered, it's usually a `UNS+S` segment, but I'm not sure if it can be something else

**Comment by Chemaclass** [comment]
I would prefer to keep support for PHP 8.0 a bit longer, until there are no more security fixes ([see orange line](https://www.php.net/supported-versions.php)). So, we could use `public const string` instead of enum. Would that be ok, @Jaimies?

**Comment by Jaimies** [comment]
That should be it

**Comment by Jaimies** [comment]
> 

I should probably use a constant there, as you said

**Comment by Jaimies** [comment]
And then we can just rollback the commit that removed the enum.

**Comment by Jaimies** [comment]
That should be the final version

**Comment by Jaimies** [comment]
Feel free to take a look and give me your feedback / criticisms, this was quite a big change for the library

File: src/MessageDataBuilder/Builder.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        if ($this->atStartOfDetailsSection($segment)) {
+            $this->setCurrentBuilder(new DetailsSectionBuilder());
+        }
+
+        if ($this->atEndOfDetailsSection($segment)) {
+            $this->setCurrentBuilder(new SimpleBuilder());
+        }
+    }
+
+    public function atStartOfDetailsSection(SegmentInterface $segment): bool
```
Comment (Chemaclass) [inline src/MessageDataBuilder/Builder.php:49]: Shouldn't this be private?
```suggestion
    private function atStartOfDetailsSection(SegmentInterface $segment): bool
```

File: src/MessageDataBuilder/Builder.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        return $segment instanceof LINLineItem
+            && !($this->currentBuilder instanceof DetailsSectionBuilder);
+    }
+
+    private function atEndOfDetailsSection(SegmentInterface $segment): bool
+    {
+        return $segment instanceof UNSSectionControl &&
+            $segment->indicatesEndOfDetailsSection();
```
Comment (Chemaclass) [inline src/MessageDataBuilder/Builder.php:58]: Can you change this to follow the convention (as you did already in the previous method)?
```suggestion
        return $segment instanceof UNSSectionControl
            && $segment->indicatesEndOfDetailsSection();
```

**Review by Chemaclass — APPROVED** [review]
Apart from two minor suggestions, looks really nice.

File: tests/Unit/MessageBuilder/MessageBuilderTest.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    private SegmentInterface $referenceSegment;
+    private SegmentInterface $dateTimeSegment;
+    private SegmentInterface $otherDateTimeSegment;
+    private SegmentInterface $lineSegment;
+    private SegmentInterface $quantitySegment;
+    private SegmentInterface $otherLineSegment;
+    private SegmentInterface $otherQuantitySegment;
+    private SegmentInterface $separatorBetweenDetailsAndSummarySegment;
+
+    public function __construct(?string $name = null, array $data = [], int|string $dataName = '')
```
Comment (Chemaclass) [inline tests/Unit/MessageBuilder/MessageBuilderTest.php:27]: Have you tried using the `setUp()` instead of the `__construct()`?

**Comment by Chemaclass** [comment]
Can you apply those three changes/suggestions, @Jaimies? Then I will gladly merge it ⭐️

File: tests/Unit/MessageBuilder/MessageBuilderTest.php (hunk 3 above)
Comment (Jaimies) [reply tests/Unit/MessageBuilder/MessageBuilderTest.php:27]: I'll give it a try

**Comment by Jaimies** [comment]
All done

**Pull request merged**